src/tls.rs

```diff
@@ -9,24 +12,107 @@ use rustls::{ClientConfig, DigitallySignedStruct, Error as TlsError, RootCertSto
 use crate::client::Result;
 use crate::Error;
 
-/// Options for tls connection.
 #[derive(Debug)]
-pub struct TlsOptions {
-    identity: Option<(Vec<CertificateDer<'static>>, PrivateKeyDer<'static>)>,
-    ca_certs: RootCertStore,
-    hostname_verification: bool,
+struct FileProvider {
+    certs: Vec<CertificateDer<'static>>,
+    key: PrivateKeyDer<'static>,
+    cert_path: PathBuf,
+    key_path: PathBuf,
+    cert_modified: SystemTime,
+    key_modified: SystemTime,
+}
+
+impl FileProvider {
+    async fn new(cert_path: PathBuf, key_path: PathBuf) -> Result<Self> {
+        let (certs, key) = load_certificates_from_files(&cert_path, &key_path).await?;
+        let (cert_modified, key_modified) = get_file_timestamps(&cert_path, &key_path).await?;
+        Ok(Self { certs, key, cert_path, key_path, cert_modified, key_modified })
+    }
+
+    async fn update_and_fetch(&mut self) -> Result<(Vec<CertificateDer<'static>>, PrivateKeyDer<'static>)> {
+        let (cert_modified, key_modified) = get_file_timestamps(&self.cert_path, &self.key_path).await?;
+        let cert_changed = cert_modified > self.cert_modified;
+        let key_changed = key_modified > self.key_modified;
+        // Refresh if both files were modified, as we want to make sure that we don't pick up a new cert/key with
+        // an old key/cert.
+        if cert_changed && key_changed {
```

Comment:
I think it is a good enhancement. But instead of `FileBased`, I would it to be more generic. I guess `TlsOptions::with_pem_reloader(self, reloader: impl Fn() -> Option<...>`) -> Self` should work. Let me take a deep look.